include/irr/asset/ICPUVirtualTexture.h

```diff
@@ -0,0 +1,337 @@
+
+            const auto& region = m_pageTable->getRegions().begin()[i];
+            const auto strides = region.getByteStrides(TexelBlockInfo(m_pageTable->getCreationParameters().format));
+
+            for (uint32_t y = 0u; y < h; ++y)
+                for (uint32_t x = 0u; x < w; ++x)
+                {
+                    uint32_t* texelptr = reinterpret_cast<uint32_t*>(bufptr + region.getByteOffset(core::vector4du32_SIMD((_addr.pgTab_x>>i) + x, (_addr.pgTab_y>>i) + y, 0u, _addr.pgTab_layer), strides));
+                    SPhysPgOffset physPgOffset = *texelptr;
+                    if (storage->physPgOffset_valid(physPgOffset))
+                    {
+                        *texelptr = SPhysPgOffset::invalid_addr;
+
+                        uint32_t addrs[2] { physPgOffset.addr&0xffffu, storage->physPgOffset_mipTailAddr(physPgOffset).addr&0xffffu };
+                        const uint32_t szs[2]{ 1u,1u };
+                        core::address_allocator_traits<phys_pg_addr_alctr_t>::multi_free_addr(storage->tileAlctr, storage->physPgOffset_hasMipTailAddr(physPgOffset) ? 2u : 1u, addrs, szs);
+                    }
+                }
+            fill.subresource.mipLevel = i;
+            fill.outRange.offset = {static_cast<uint32_t>(_addr.pgTab_x>>i),static_cast<uint32_t>(_addr.pgTab_y>>i),0u};
+            fill.outRange.extent = {w,h,1u};
+            CFillImageFilter::execute(&fill);
+        }
+
+        //free entries in page table
+        if (!base_t::free(_addr, _img, _subres))
+            return false;
+        //in case when pgtab layer has no allocations, free it for use by another format
+        if ((*m_pageTableLayerAllocators)[_addr.pgTab_layer].get_allocated_size()==0u)
+        {
+            (*m_pageTableLayerAllocators)[_addr.pgTab_layer].reset();//why actually do i need to reset it if its allocated size is 0 anyway?
+            (*m_layerToViewIndexMapping)[_addr.pgTab_layer] = ~0u;
+            storage->removeLayerAssignment(_addr.pgTab_layer);
+            m_freePageTableLayerIDs.push(_addr.pgTab_layer);
+        }
+
+        return true;
+    }
+
+    core::smart_refctd_ptr<ICPUImageView> createPageTableView() const override
+    {
+        ICPUImageView::SCreationParams params;
+        params.flags = static_cast<ICPUImageView::E_CREATE_FLAGS>(0);
+        params.format = m_pageTable->getCreationParameters().format;
+        params.subresourceRange.aspectMask = static_cast<IImage::E_ASPECT_FLAGS>(0);
+        params.subresourceRange.baseArrayLayer = 0u;
+        params.subresourceRange.layerCount = m_pageTable->getCreationParameters().arrayLayers;
+        params.subresourceRange.baseMipLevel = 0u;
+        params.subresourceRange.levelCount = m_pageTable->getCreationParameters().mipLevels;
+        params.viewType = IImageView<ICPUImage>::ET_2D_ARRAY;
+        params.image = m_pageTable;
+
+        return ICPUImageView::create(std::move(params));
+    }
```

Comment:
See 
https://github.com/buildaworldnet/IrrlichtBAW/pull/474/files?file-filters%5B%5D=.cpp&file-filters%5B%5D=.h&file-filters%5B%5D=No+extension&hide-deleted-files=true#r417876361
and
https://github.com/buildaworldnet/IrrlichtBAW/pull/474/files?file-filters%5B%5D=.cpp&file-filters%5B%5D=.h&file-filters%5B%5D=No+extension&hide-deleted-files=true#r417875791